openshift/default-catalog-consistency/pkg/extract/extract.go

```diff
@@ -158,23 +160,33 @@ func extractLayers(ctx context.Context, layoutPath, fsPath, tag string) error {
 			return fmt.Errorf("decompress blob %d: %w", i, err)
 		}
 
-		// To avoid permission errors faced when extracting the layers
-		mask := os.FileMode(0755)
-		opts := &archive.TarOptions{
-			ForceMask: &mask,
-			// Required to avoid permission errors when extracting the layers in the OCP CI environment.
-			// extract filesystem: apply layer 0: lchown /tmp/.../afs: operation not permitted
-			NoLchown: true,
-		}
-
-		_, err = archive.ApplyUncompressedLayer(fsPath, decompress, opts)
-		decompress.Close()
-		rc.Close()
-
+		err = func() error {
+			defer decompress.Close()
+			defer rc.Close()
+
+			_, err := archive.Apply(ctx, fsPath, decompress, archive.WithFilter(func(hdr *tar.Header) (bool, error) {
+				// Clean up extended headers and enforce safe permissions
+				hdr.PAXRecords = nil
+				hdr.Xattrs = nil
+				hdr.Uid = os.Getuid()
+				hdr.Gid = os.Getgid()
+				if hdr.FileInfo().IsDir() {
+					hdr.Mode = 0700
+				} else {
+					hdr.Mode = 0600
+				}
+				return true, nil
+			}))
+			if err != nil {
+				return fmt.Errorf("apply layer %d: %w", i, err)
+			}
+			return nil
+		}()
 		if err != nil {
-			return fmt.Errorf("apply layer %d: %w", i, err)
+			return fmt.Errorf("decompress layer %d: %w", i, err)
```

Comment:
```suggestion
			return fmt.Errorf("apply layer %d: %w", i, err)
```

Reply:
Good call. Most important was to ensure that it would work with the OCP CI.
It worked 🎉 
Doing a follow up for this small cleanups